browser/ui/webui/settings/brave_extensions_manifest_v2_installer.cc

```diff
@@ -0,0 +1,222 @@
+  CHECK(url.is_valid());
+  return url;
+}
+
+GURL GetCrxDownloadUrl(const base::Value::Dict& update_manifest,
+                       std::string_view extension_id) {
+  const auto* gupdate = update_manifest.FindDict("gupdate");
+  if (!gupdate) {
+    return GURL();
+  }
+  const auto* app_list = gupdate->FindList("app");
+  if (!app_list) {
+    return GURL();
+  }
+
+  for (const auto& appv : *app_list) {
+    if (!appv.is_dict()) {
+      continue;
+    }
+    const auto& app = appv.GetDict();
+
+    const auto* id = app.FindString("appid");
+    if (!id || *id != extension_id) {
+      continue;
+    }
+
+    const auto* update_check = app.FindDict("updatecheck");
+    if (!update_check) {
+      break;
+    }
+    const auto* codebase = update_check->FindString("codebase");
+    if (!codebase) {
+      break;
+    }
+    const GURL url(*codebase);
+    if (!url.is_valid() || !brave::ShouldAddBraveServicesKeyHeader(url)) {
+      break;
+    }
+    return url;
+  }
+
+  return GURL();
+}
+
+}  // namespace
+
+ExtensionManifestV2Installer::ExtensionManifestV2Installer(
+    const std::string& extension_id,
+    content::WebContents* web_contents,
+    extensions::WebstoreInstallWithPrompt::Callback callback)
+    : extension_id_(extension_id),
+      web_contents_(web_contents->GetWeakPtr()),
+      callback_(std::move(callback)) {
+  CHECK(IsKnownMV2Extension(extension_id));
+  auto* profile =
+      Profile::FromBrowserContext(web_contents_->GetBrowserContext());
+  url_loader_factory_ = profile->GetURLLoaderFactory();
+}
+
+ExtensionManifestV2Installer::~ExtensionManifestV2Installer() = default;
+
+void ExtensionManifestV2Installer::BeginInstall() {
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = GetUpdaterExtensionDownloadUrl(extension_id_);
+  request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  request->headers.SetHeader("Content-Type", "application/json");
+  request->headers.SetHeader(kBraveServicesKeyHeader,
+                             BUILDFLAG(BRAVE_SERVICES_KEY));
+
+  constexpr auto kExtensionsMV2ManifestAnnotationTag =
+      net::DefineNetworkTrafficAnnotation("extensions_mv2_request", R"(
+    semantics {
+      sender: "Extension Manifest V2 Installer"
+      description:
+        "In response to this request Brave backend returns an JSON file
+        with update/download response."
+      trigger:
+        "The user enables MV2 extension on the settings page"
+      destination: BRAVE_OWNED_SERVICE
+    }
+    policy {
+      cookies_allowed: NO
+    })");
+
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(request), kExtensionsMV2ManifestAnnotationTag);
+  url_loader_->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
+      url_loader_factory_.get(),
+      base::BindOnce(&ExtensionManifestV2Installer::OnUpdateManifestResponse,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void ExtensionManifestV2Installer::OnUpdateManifestResponse(
```

Comment:
why can't we use `ExtensionDownloader`?

Reply:
This class was needed primarily because we don’t want to sign third-party extensions with our publisher key - otherwise, we could have used `WebstoreInstaller` directly. But now, after taking another look, it seems easier to adapt `CrxInstaller` within `WebstoreInstaller` rather than creating a custom loader.

Reply:
I'm wrong, `WebstoreInstaller` relies on the web store API. I'll try to use `ExtensionDownloader`

Reply:
`ExtensionDownloader` uses protocol version 2.0, but our backend supports 3.1 and higher. We could add support for protocol 2.0 - for our use case, the response is essentially the same as 3.1. We would only need to:
1) Add the namespace `xmlns="http://www.google.com/update2/response"`
2) Change the `protocol` attribute in the `gupdate` element.

Updates for extensions currently use protocol 4.0.

Or we can patch the response parser and remove checks for `xmlns` and `protocol` version.